File: gems/persona/lib/persona/account_link.rb
module Persona
  # Binds a verified OIDC identity (provider + subject) to a persona.
  #
  # Structurally the same event as redeeming a recovery code: the subject
  # identifies a "holder" — the persona already bound to it, if any — and the
  # acting browser is either adopted into that holder or has the subject
  # linked to it. When the browser is already a *different* joined persona
  # with contributions, the two are folded together via the store's merge,
  # gated by a confirmation preview.
  #
  # Callers pass the browser's agent_uid so an anonymous browser can be
  # bound; current_user (resolved from the X-Agent-Id header or the session
  # cookie, depending on transport) is the persona the browser is currently
  # acting as, or nil when still anonymous.
  #
  # All persistence goes through a storage port so the decision logic stays
  # host-agnostic. The port is an object implementing:
  #
  #   holder_for(provider:, subject:)            -> user | nil
  #   within_transaction { ... }                 atomic scope; yields
  #   create_guest!(agent_uid:, user_agent:)     -> user, already bound to
  #                                              agent_uid, with the app's
  #                                              usual join side effects run
  #   add_account_binding!(user, provider:, subject:)
  #   agent_binding?(user, agent_uid:)           -> boolean
  #   add_agent_binding!(user, agent_uid:, user_agent:)
  #   merge_preview(source:, target:)            -> opaque preview object,
  #                                              MUST NOT change data
  #   merge!(source:, target:)                   move all bindings and domain
  #                                              records from source to
  #                                              target, then retire source
  #
  # User objects are the host's own; the only requirement is a stable #id.
  # Consumers register their implementation via
  # Persona.config.account_link_store (an ActiveRecord example lives in
  # examples/ at the repository root).
  module AccountLink
    # agent_uid is set only when the browser must persist a (new) one via
    # setAgentId — i.e. the anonymous cases. merge_preview is present only
    # when confirmation is required, in which case NOTHING was changed.
    Result = Struct.new(:user, :agent_uid, :merged, :merge_preview, keyword_init: true)

    module_function

    def perform(identity:, current_user:, agent_uid:, user_agent: nil, confirm_merge: false,
                store: Persona.config.account_link_store)
      raise ConfigurationError, 'Persona.config.account_link_store is not configured — ' \
                                'provide an object implementing the AccountLink storage port' if store.nil?

      holder = holder_for(identity, store: store)

      # Conflict peek: the browser is joined as someone other than the
      # subject's current holder. Return the preview without touching data so
      # the user can still cancel; the client re-sends with confirm_merge.
      if current_user && holder && current_user.id != holder.id && !confirm_merge
        return Result.new(
          user: current_user,
          agent_uid: nil,
          merged: false,
          merge_preview: store.merge_preview(source: current_user, target: holder),
        )
      end

      store.within_transaction do
        # Subject not yet linked to any persona: bind it to the acting
        # persona, creating a fresh guest first if the browser is anonymous.
        if holder.nil?
          holder = current_user || store.create_guest!(agent_uid: agent_uid, user_agent: user_agent)
          store.add_account_binding!(holder, provider: identity.provider, subject: identity.subject)
        end

        merged = false
        if current_user && current_user.id != holder.id
          store.merge!(source: current_user, target: holder)
          merged = true
        end

        # Ensure this browser's agent_uid resolves to the holder. For a joined
        # browser that was just merged, merge! already moved its binding; for
        # an anonymous browser adopting an existing holder, add it now.
        unless store.agent_binding?(holder, agent_uid: agent_uid)
          store.add_agent_binding!(holder, agent_uid: agent_uid, user_agent: user_agent)
        end

        Result.new(
          user: holder,
          agent_uid: current_user ? nil : agent_uid,
          merged: merged,
          merge_preview: nil,
        )
      end
    end

    def holder_for(identity, store: Persona.config.account_link_store)
      store.holder_for(provider: identity.provider, subject: identity.subject)
    end
  end
end

File: packages/server-core/src/account_link.ts
// The claim decision table (docs/spec §6): binds a verified identity
// (provider + subject) — or any claim subject — to a persona. Mirrors the
// Ruby implementation (gems/persona/lib/persona/account_link.rb); both are
// checked against the same conformance vectors.
//
// All persistence goes through a storage port so the decision logic stays
// host-agnostic. User objects are the host's own; the only requirement is a
// stable `id`.

export type ClaimIdentity = {
  provider: string;
  subject: string;
};

export type AccountLinkStore<User extends { id: unknown }> = {
  // The persona currently bound to (provider, subject), if any.
  holderFor(provider: string, subject: string): Promise<User | null>;
  // Atomic scope for the mutating half of the decision table.
  withinTransaction<T>(fn: () => Promise<T>): Promise<T>;
  // A fresh persona already bound to agentUid, with the application's usual
  // join side effects run.
  createGuest(agentUid: string, userAgent?: string): Promise<User>;
  addAccountBinding(user: User, provider: string, subject: string): Promise<void>;
  hasAgentBinding(user: User, agentUid: string): Promise<boolean>;
  addAgentBinding(user: User, agentUid: string, userAgent?: string): Promise<void>;
  // MUST NOT change data (P-6). The preview value is opaque to the kit.
  mergePreview(source: User, target: User): Promise<unknown>;
  // Move all bindings and domain records from source to target, then retire
  // source (P-13).
  merge(source: User, target: User): Promise<void>;
};

export type AccountLinkResult<User> = {
  user: User;
  // Set only when the browser must persist a (new) agent_uid — i.e. the
  // anonymous cases (the credential-echo exception, H-3 / C-3).
  agentUid: string | null;
  merged: boolean;
  // Present only when confirmation is required, in which case NOTHING was
  // changed (P-6).
  mergePreview: unknown | null;
};

export const performAccountLink = async <User extends { id: unknown }>(opts: {
  identity: ClaimIdentity;
  currentUser: User | null;
  agentUid: string;
  userAgent?: string;
  confirmMerge?: boolean;
  store: AccountLinkStore<User>;
}): Promise<AccountLinkResult<User>> => {
  const { identity, currentUser, agentUid, userAgent, confirmMerge = false, store } = opts;

  let holder = await store.holderFor(identity.provider, identity.subject);

  // Conflict peek: the browser is joined as someone other than the subject's
  // current holder. Return the preview without touching data so the user can
  // still cancel; the client re-sends with confirmMerge.
  if (currentUser && holder && currentUser.id !== holder.id && !confirmMerge) {
    return {
      user: currentUser,
      agentUid: null,
      merged: false,
      mergePreview: await store.mergePreview(currentUser, holder),
    };
  }

  return store.withinTransaction(async () => {
    // Subject not yet linked to any persona: bind it to the acting persona,
    // creating a fresh guest first if the browser is anonymous.
    if (!holder) {
      holder = currentUser ?? (await store.createGuest(agentUid, userAgent));
      await store.addAccountBinding(holder, identity.provider, identity.subject);
    }

    let merged = false;
    if (currentUser && currentUser.id !== holder.id) {
      await store.merge(currentUser, holder);
      merged = true;
    }

    // Ensure this browser's agent_uid resolves to the holder. For a joined
    // browser that was just merged, merge already moved its binding; for an
    // anonymous browser adopting an existing holder, add it now.
    if (!(await store.hasAgentBinding(holder, agentUid))) {
      await store.addAgentBinding(holder, agentUid, userAgent);
    }

    return {
      user: holder,
      agentUid: currentUser ? null : agentUid,
      merged,
      mergePreview: null,
    };
  });
};
